### PR TITLE
[Backpack]: improve the error message in case of a mutually recursive unit

### PR DESCRIPTION
**Pull request by ak3n**
This PR (reincarnation of #7229) improves the error message in case when the user adds a dependency with signature to the implementation:

```
library interface
  hs-source-dirs: interface
  signatures: A
  build-depends: base

library impl
  hs-source-dirs: impl
  exposed-modules: ImplA
  reexported-modules: ImplA as A
  build-depends: base, interface

executable exe
  main-is: Main.hs
  build-depends: base, impl, interface
```

The error message was:

```
Error:
    Unsupported mutually recursive unit identifier
    In the stanza 'executable exe'
    In the inplace package 'example-0.1.0.0'
```

The improved error message:

```
Error:
    Cannot instantiate requirement  'ImplA'
    Check that "build-depends:" doesn't include library with signatures:  'A'
    as this creates a cyclic dependency, which GHC does not support.
    In the stanza 'executable exe'
    In the inplace package 'example-0.1.0.0'
```

which is almost the same as in https://github.com/haskell/cabal/blob/3cd9fded57805ab19ee15474051ec87c7d87cd39/Cabal/src/Distribution/Backpack/MixLink.hs#L70

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by ulysses4ever** [comment]
FYI change logs are explained here: https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog a tip: this is referenced from the PR template.

File: Cabal/src/Distribution/Backpack/UnifyM.hs
Comment (Mikolaj) [inline Cabal/src/Distribution/Backpack/UnifyM.hs:417]: Could this and the previous occurrence be factored out? That would make it apparent in the code that a similar thing is outputted in two similar cases.

File: Cabal/src/Distribution/Backpack/LinkedComponent.hs
Comment (Mikolaj) [inline Cabal/src/Distribution/Backpack/LinkedComponent.hs:223]: I'm not a native speaker, but would "Ensure" instead of "Check" be more precise? Then "that" may be skipped. I'd probably also say "any library". BTW, is it "library" or "package" or "component" or what? How does https://cabal.readthedocs.io/en/latest/ name the things after `build-depends`?

BTW, are the double spaces intentional (the extra one seems to be coming from `<+>`)?

**Comment by Mikolaj** [comment]
The CI is fixed (again), so let me rebase to fix CI results.

**Comment by Mikolaj** [comment]
@mergify rebase

**Review by Mikolaj — APPROVED** [review]
LGTM. Thank you. Now we wait for the second review.

File: Cabal/src/Distribution/Backpack/UnifyM.hs
Comment (Ericson2314) [inline Cabal/src/Distribution/Backpack/UnifyM.hs:404]: why not?
```suggestion
convertModuleU' :: MooEnv -> ModuleU s -> UnifyM s (Either [ModuleName] OpenModule)
```

**Review by Ericson2314 — CHANGES_REQUESTED** [review]
Stepping back a bit, either *both* functions or *neither* function should return either.

TBH since everyone is handling the error in the same way, I think it is easiest if *neither* function returns `Either`, and we just call the error message function right away.

If we do want to go with the `Either` route I think it would better to not use `UnifyM` to demonstrate in the types that error handling is deferred to the caller, but that is more work and could fail because some other called function. Hence why I recommend just going with the "no-`Either`" route for now.

**Comment by Ericson2314** [comment]
@ak3n I don't super mind that, but if you feel especially motivated you could changed `UnifM` so that it used a structured errors like GHC rather than a list of doc-strings (if the other reviewers approve).

Then it is possible for the caller to add-context without cluttering the callee. For example, could make something like outer "stack frames" for a "back trace" showing the component id and module name!

**Comment by Ericson2314** [comment]
@ak3n Aabsolutely! I just brought it up as a way to make the extra argument temporary so we don't feel bad about it.

**Review by Ericson2314 — APPROVED** [review]
Thanks!!

**Comment by Ericson2314** [comment]
Oh yes, sorry I didn't do that earlier!

**Comment by Mikolaj** [comment]
> @Mikolaj @ulysses4ever Sorry, can you merge the PR please?

Yes, of course. Even better, you can merge yourself, if you accept the invitation I've just sent you:

https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions

Thanks a lot of the PR!

**Pull request merged**
